kawin/Diffusion.py

```diff
@@ -0,0 +1,825 @@
+
+        # 1 / ((1 / mPhi - mAlpha) + 1 / (3mAlpha)) = 3mAlpha * (mPhi - mAlpha) / (2mAlpha + mPhi)
+        Ak = 3 * maxMob * (mob - maxMob) / (2*maxMob + mob)
+        Ak = Ak * self.p[:,np.newaxis]
+        Ak = np.sum(Ak, axis=0)
+        avgMob = maxMob + Ak / (1 - Ak / (3*maxMob))
+        return avgMob
+
+    def hashin_shtrikmanLower(self, xarray):
+        '''
+        Lower hashin shtrikman bounds for average mobility
+
+        Returns
+        -------
+        (e, N) mobility array - e is number of elements, N is number of nodes
+        '''
+        #self.p                                 #(p,N)
+        mob = self.getMobility(xarray)          #(p,e+1,N)
+        minMob = np.amin(mob, axis=0)           #(e+1,N)
+
+        #This prevents an infinite mobility which could cause the time interval to be 0
+        minMob[minMob == np.inf] = 0
+
+        # 1 / ((1 / mPhi - mAlpha) + 1 / (3mAlpha)) = 3mAlpha * (mPhi - mAlpha) / (2mAlpha + mPhi)
+        Ak = 3 * minMob * (mob - minMob) / (2*minMob + mob)
+
+        Ak = Ak * self.p[:,np.newaxis]
+        Ak = np.sum(Ak, axis=0)
+        avgMob = minMob + Ak / (1 - Ak / (3*minMob))
+        return avgMob
+
+    def getFluxes(self):
+        self.p = self.updateCompSets(self.x)
+        #for i in range(self.N):
+        #    print(i, self.compSets[i], self.p[:,i], self.x[:,i])
+
+        #Get average mobility between nodes
+        avgMob = self.mobilityFunction(self.x)
+        avgMob = 0.5 * (avgMob[:,1:] + avgMob[:,:-1])
+
+        #Composition between nodes
+        avgX = 0.5 * (self.x[:,1:] + self.x[:,:-1])
+        avgX = np.concatenate(([1-np.sum(avgX, axis=0)], avgX), axis=0)
+
+        #Chemical potential gradient
+        dmudz = (self.mu[:,1:] - self.mu[:,:-1]) / self.dz
+
+        #Composition gradient (we need to calculate gradient for reference element)
+        dxdz = (self.x[:,1:] - self.x[:,:-1]) / self.dz
+        dxdz = np.concatenate(([0-np.sum(dxdz, axis=0)], dxdz), axis=0)
+
+        # J = -M * dmu/dz
+        # Ideal contribution: J_id = -eps * M*R*T / x * dx/dz
+        fluxes = np.zeros((len(self.elements)+1, self.N-1))
+        fluxes = -avgMob * dmudz
+        nonzeroComp = avgX != 0
+        fluxes[nonzeroComp] += -self.eps * avgMob[nonzeroComp] * 8.314 * self.T * dxdz[nonzeroComp] / avgX[nonzeroComp]
+
+        #Flux in a volume fixed frame: J_vi = J_i - x_i * sum(J_j)
+        vfluxes = np.zeros((len(self.elements), self.N+1))
+        vfluxes[:,1:-1] = fluxes[1:,:] - avgX[1:,:] * np.sum(fluxes, axis=0)
+
+        #Boundary conditions
+        for e in range(len(self.elements)):
+            vfluxes[e,0] = self.LBCvalue[e] if self.LBC[e] == self.FLUX else vfluxes[e,1]
+            vfluxes[e,-1] = self.RBCvalue[e] if self.RBC[e] == self.FLUX else vfluxes[e,-2]
+
+        #Time increment
+        #HOW DO WE DO THIS!?!?!?
+        dJ = np.abs(vfluxes[:,1:] - vfluxes[:,:-1]) / self.dz
+        dt = self.maxCompositionChange / np.amax(dJ[dJ!=0])
+        #print(dt)
```

Comment:
Can you say a little more about what's causing trouble here?

Reply:
This is actually fixed for the most part. That comment was there when I originally started the homogenization model and I forgot to remove it.

The time increment is found by making sure the change in composition for each volume element is less than or equal to `maxCompositionChange`